File: OOD C++/Assignment5/Composite.cpp
#include "Polyhedron.h"
#include "Composite.h"

//------------------------------------------------------------------------------
Composite::Composite()
    :Polyhedron("Composite")
{
}

//------------------------------------------------------------------------------
/**
 * @todo write this function
 */
Composite::Composite(const Composite& src)
    :Polyhedron("Composite")
{
    // Perform a deep copy... maybe the _add_ method can help...
    for(const Polyhedron* p : src.allPolyhedra){
        this->add(p);
    }
}

//------------------------------------------------------------------------------
/**
 * @todo write this function
 */
Composite::~Composite()
{
    // Delete each component polyhedra
    for(Polyhedron* p : this->allPolyhedra){
        delete p;
    }
}

//------------------------------------------------------------------------------
void Composite::read(std::istream& ins){
    int numPolyhedra;

    ins >> numPolyhedra;

    allPolyhedra.resize(numPolyhedra);

    for (int i = 0; i < numPolyhedra; i++) {
        ins >> allPolyhedra[i];

        boundingBox.merge(allPolyhedra[i]->getBoundingBox());
    }
}

//------------------------------------------------------------------------------
/**
 * @todo write this function
 */
void Composite::display(std::ostream& outs) const
{
    Polyhedron::display(outs);

    outs << allPolyhedra.size() << " polyhedra" << "\n";

    // Loop through all component polyhedra and
    // print (display) them
    for(Polyhedron* p : this->allPolyhedra){
        outs << "  " << *p << "\n";
    }
}

//------------------------------------------------------------------------------
/**
 * @todo write this function
 */
void Composite::scale(double scalingFactor)
{
    // Loop through all polyhedra and scale them
    for(Polyhedron* p : this->allPolyhedra){
        p->scale(scalingFactor);
    }

    // Do not forget the bounding box... after the loop
    boundingBox.scale(scalingFactor);
}

//------------------------------------------------------------------------------
Composite& Composite::operator=(Composite rhs)
{
    swap(*this, rhs);
    return *this;
}

//------------------------------------------------------------------------------
Composite::iterator Composite::begin()
{
    return allPolyhedra.begin();
}

//------------------------------------------------------------------------------
Composite::iterator Composite::end()
{
    return allPolyhedra.end();
}

//------------------------------------------------------------------------------
Composite::const_iterator Composite::begin() const
{
    return allPolyhedra.begin();
}

//------------------------------------------------------------------------------
Composite::const_iterator Composite::end() const
{
    return allPolyhedra.end();
}

//------------------------------------------------------------------------------
/**
 * @todo write this function
 */
void Composite::add(const Polyhedron* toAdd)
{
    // Add one new polyhedra and _merge_ its boundingBox with _this->boundingBox_
    allPolyhedra.push_back(toAdd->clone());
    boundingBox.merge(toAdd->getBoundingBox());
}

//------------------------------------------------------------------------------
void swap(Composite& lhs, Composite& rhs)
{
    using std::swap;

    std::swap(lhs.allPolyhedra, rhs.allPolyhedra);

    swap(lhs.boundingBox, rhs.boundingBox);
}

File: Object-Orient-Program-and-Design-Java/Assignment8/src/main/java/polyhedra/Composite.java
package polyhedra;

import java.util.List;
import java.util.Vector;
import java.util.Iterator;

/**
 * A Composite Polyhedron. Polyhedra of this type are built from
 * other Polyhedra. This,in theory, can include Composite objects
 * composed of other (nested) Composite objects.
 *
 */
public class Composite
    implements Polyhedron, Iterable<Polyhedron>
{
    /**
     * Cached (i.e., precomputed and updated) BoundingBox.
     *
     * Computing the boundingbox for a Composite Polyhedron is non-trivial
     * (i.e., requires a loop and has O(n) temperoral complexity).
     */
    private BoundingBox theBox;

    /**
     * Collection of polyhedra of which this composite polyhedron is composed.
     */
    private List<Polyhedron> allPolyhedra;

    /**
     * Default Composite Constructor.
     */
    public Composite()
    {
        this.allPolyhedra = new Vector<Polyhedron>();
        this.theBox = new BoundingBox();
    }

    @Override
    public String getType()
    {
        // Replace the return line
        return "Composite";
    }

    @Override
    public BoundingBox getBoundingBox()
    {
        return theBox;
    }

    /**
     * Add a cloned copy of a Polyhedron to the `Composite` collection.
     *
     * @param toAdd is cloned and the copy is added
     */
    public void add(Polyhedron toAdd)
    {
        // Write this function.
        allPolyhedra.add(toAdd.clone());
        theBox.merge(toAdd.getBoundingBox());
    }

    /**
     * Iterate through all sub-polyhedra, scale them, and update all
     * bounding boxes.
     *
     * @param scalingFactor scaling factor
     */
    public void scale(double scalingFactor)
    {
        // Write this function.
        for(Polyhedron p : this.allPolyhedra) {
            p.scale(scalingFactor);
        }

        theBox.scale(scalingFactor);
    }

    /**
     * Retrive the number of Polyhedra.
     *
     * @return the number of Polyhedra that comprise this Composite object
     */
    public int size()
    {
        return allPolyhedra.size();
    }

    @Override
    public Iterator<Polyhedron> iterator()
    {
        return allPolyhedra.iterator();
    }

    @Override
    public Polyhedron clone()
    {
        Composite aCopy = new Composite();

        // A loop might be helpful to 'add' each entry from this.allPolyhedra
        for(Polyhedron poly : this.allPolyhedra){
            aCopy.add(poly);
        }

        return aCopy;
    }

    @Override
    public boolean equals(Object rhs)
    {
        if (!(rhs instanceof Composite)) {
            return false;
        }

        Composite rhsComp = (Composite) rhs;

        // List.equals compares each pair in order (magic)!
        return this.allPolyhedra.equals(rhsComp.allPolyhedra);
    }

    @Override
    public int hashCode()
    {
        return this.getType().hashCode()
             + this.allPolyhedra.hashCode();
    }

    /**
     * "Print" all polyhedra.
     *
     * @return String containing all component _Polyhedra_ objects
     */
    @Override
    public String toString()
    {
        StringBuilder bld = new StringBuilder();

        bld.append(Polyhedron.super.toStringHelper());
        bld.append(String.format("%d polyhedra%n", this.size()));

        // Maybe a loop can help...
        for(Polyhedron poly : this.allPolyhedra) {
            bld.append("  " + poly + "\n");
        }

        return bld.toString();
    }

    //--------------------------------------------------------------------------
    @Override
    public boolean isComplex()
    {
        // Is the return corrrect?
        return true;
    }

    @Override
    public boolean isSimple()
    {
        // Is the return corrrect?
        return false;
    }
}

